Lab10

nicko-lee opened the pull request:
Fixes #641 

Review of colleague's PR #636 

#### Changes proposed in this PR:

#### Implement lab10 functionality + write tests
- Create puppyHandler which is essentially a REST wrapper around our Storer interface
- Which exposes the functionality we built over the past few labs over a REST API
- Expand flag for user to specify storer type as well as other config settings like port number and file to load data from
- Learn how to use net/http/httptest package for testing REST APIs

Commit:
Implement lab10 features + tests

Create Rest API wrapper around features previously implemented
to date

Commit:
Fix golangci-lint issues + getting test coverage to 100%

Commit:
Update README.md file to incorporate lab10 relevant info

Commit:
Fix minor error in README

anzboi commented on 10_rest/nickolee/pkg/puppy/rest/rest.go:
```diff
@@ -0,0 +1,139 @@
+// a special guy who wraps around Storer types and gives them extra abilities to now handle API requests
+// thus enhancing their storing abilities significantly!
+type PuppyHandler interface {
+	handlePost(w http.ResponseWriter, r *http.Request)
```
Your linting issue comes from the fact that `SetupRoutes` takes a `PuppyHandlerAndStorer`, not a `PuppyHandler`.

This means `PuppyHandler` really isn't used at all, and cannot be used outside this package, so is effectively dead code.

To fix, your interface methods need to be exported, `SetupRoutes` needs to take the interface.

IMO PuppyHandler doesn't even need to be an interface, and can be the struct with the store. Some people disagree with this though, you can go either way.

In the end, you would have a package that looks like this

```go
type PuppyHandler interface {
    HandlePost(args) // notice it is exported
    // other methods
}

type PuppyHandlerAndStore struct {
  Store puppy.Storer
}

// implement PuppyHandler methods for puppyHandlerAndStore

func SetupRoutes(r chi.Router, ph PuppyHandler) {
  r.Get("path", ph.HandleGet)
  // other routes
}
```

Commit:
Fix final lint error

Unwittingly had dead code in the form of an unused interface. In the end
I opted to pass a struct into SetupRoutes() instead of passing an interface

Commit:
Fix super pesky race condition issues to pass CI

Had to use some workarounds

anzboi commented on 10_rest/nickolee/README.md:
```diff
@@ -0,0 +1,103 @@
+
+To build the project, run the following command while you are in the root of the project folder:
+
+`go build ./...`
```
Usually we like to be a little more specific as to what we are building.
```bash
go build ./cmd/puppy-server
```

nicko-lee replied:
ah i see so if my project has got more than one binary - go build ./... will build all binaries. In this case it won't make a difference but that is clear. Got it. Being clear about exactly what i am building

nicko-lee replied:
but if do go install that would be go install ./... usually because i want all the binaries

anzboi commented on 10_rest/nickolee/README.md:
```diff
@@ -0,0 +1,103 @@
+Note that there are three flags which work with main.go:
+
+1. **-d,--data:** if you wish to specify a .json file to load in seed data for the database, pass it as a value to this flag. Note that if no file is specified then a default file is used.
+2. **-s, --store:** either _map_ or _sync_ Note that if not set then the value defaults to _map_
```
Missing period
```suggestion
2. **-s, --store:** either _map_ or _sync_. Note that if not set then the value defaults to _map_
```

anzboi commented on 10_rest/nickolee/README.md:
```diff
@@ -0,0 +1,103 @@
+
+Note that there are three flags which work with main.go:
+
+1. **-d,--data:** if you wish to specify a .json file to load in seed data for the database, pass it as a value to this flag. Note that if no file is specified then a default file is used.
```
Should not use a default file, should just not load any initial data.
Your project could provide an example a user can use to try out the functionality.

nicko-lee replied:
Got it. Normally it should default to no file. If I don't give it a file it should default to having no puppies in it.

anzboi commented on 10_rest/nickolee/README.md:
```diff
@@ -0,0 +1,103 @@
+
+1. **-d,--data:** if you wish to specify a .json file to load in seed data for the database, pass it as a value to this flag. Note that if no file is specified then a default file is used.
+2. **-s, --store:** either _map_ or _sync_ Note that if not set then the value defaults to _map_
+3. **-p, --port:** if set then an **API** server is started on the user specified port. Must be a valid port number between 0 and 65535
```
What if it is not set?

nicko-lee replied:
It defaults to 7777

nicko-lee replied:
Will add it to the README

nicko-lee replied:
defaults to port 7777. Have added that to the readme

anzboi commented on 10_rest/nickolee/README.md:
```diff
@@ -0,0 +1,103 @@
+-   Clone this project outside your `$GOPATH` to enable [Go Modules](https://github.com/golang/go/wiki/Modules)
+-   Install `golangci-lint` according to [instructions](https://github.com/golangci/golangci-lint#local-installation)
+
+## How to build this project?
```
These headings are much bigger than normal
```
## Build

## Run

## Test
```

nicko-lee replied:
Got it

anzboi commented on 10_rest/nickolee/README.md:
```diff
@@ -0,0 +1,103 @@
+## API Spec
+
+### Create puppies
+ Endpoint: {host:port}/api/puppy/
```
A few things here...

This doesn't print as well on github as you might hope.

You never include any of `scheme://host:port` in the api spec. This just tells your machine where to find (host:port) and how to connect (scheme) to the physical server serving the api.

Moreover, normally an API spec would be documented in the form of a `swagger.yaml` file. We don't require that you learn swagger. If you wanted to keep this here, you could do it in a form like the following

```
### POST /api/puppy

Adds a new puppy to the store, returns the stored puppy with a generated id.

Example:
    curl -X POST http://localhost:7777/api/puppy/ \
      -d '{
        "breed": "Iron Dog",
        "colour": "Red",
        "value": 9500
      }'

returns:
      -d '{
        "id": 1
        "breed": "Iron Dog",
        "colour": "Red",
        "value": 9500
      }'

### GET /api/puppy/{id}

...
```

nicko-lee replied:
Have changed this

anzboi commented on 10_rest/nickolee/cmd/puppy-server/main.go:
```diff
@@ -0,0 +1,122 @@
+	args = os.Args[1:]
+	// this loads your file once parseFlags() is run
+	// these are methods too prepare a variable. Basically telling kingpin how to read the flag from cli
+	fileName = kingpin.Flag("data", "This flag supplies a puppy JSON file name").
```
While the input is a filename, kingpin is returning a file struct. This variable should probably be named so that is suggest it's not merely a `string`, eg:
`datafile`

nicko-lee replied:
have changed the logic around this to actually return a *string so that I can check if no file is provided without kingpin doing the checks and panicking me out

anzboi replied:
Fair enough

anzboi commented on 10_rest/nickolee/cmd/puppy-server/main.go:
```diff
@@ -0,0 +1,122 @@
+
+func main() {
+	// parse flags/check if user even provided any
+	cliConfig, err := parseFlags(args)
```
I wouldn't bother with a separate parseFlags function.
```
func main() {
  kingpin.CommandLine.Parse(args)
  ...
  // use the parsed flag values directly
  puppies, err := readFileAndUnmarshalPuppies(*datafile)
}
```
I don't think this will affect coverage

WARNING: DO NOT use the flag variables outside of main, they are globals and WILL destroy your tests on race conditions.

anzboi replied:
It's fine if you choose not to do this, not a big deal (in fact, for more complicated services, usually the first thing you do is read any input into config).

anzboi commented on 10_rest/nickolee/cmd/puppy-server/main.go:
```diff
@@ -0,0 +1,122 @@
+	}
+
+	// read JSON from file + parse/unmarshal the json and we now have []puppy (Go objects)
+	puppies, err := readFileAndUnmarshalPuppies(cliConfig.file)
```
Move this block closer to where the puppies are loaded into the store, It's a little weird doing other stuff in the middle that are completely unrelated to these two tasks.

anzboi replied:
I would still like you to move this down. Basically, the current flow is
1. read data file
2. create store
3. create handler
4. create address
5. load initial data into store
6. setup router
7. serve


You can move the first one down to just before 5, and also move 3 to before 6 or 7, so it looks more like
1. create store
2. create handler
3. read data file
4. load initial data into store
5. setup router
6. create address
7. serve

nicko-lee replied:
ah i get what u mean now. Thanks for clarifying that! Makes sense to me now and have moved the block to a more logical order

anzboi commented on 10_rest/nickolee/cmd/puppy-server/main.go:
```diff
@@ -0,0 +1,122 @@
+		pup := pup // to avoid scopelint error: Using a reference for the variable on range scope `pup`
+		createdPuppy, _ := phs.Storage.CreatePuppy(&pup)
+		fmt.Fprintf(out, "Puppy with ID %d has been created\n", createdPuppy)
+		retrievedPuppy, _ := phs.Storage.ReadPuppy(createdPuppy)
```
Don't bother reading the puppy back out, this just produces extra noise that isn't necessary.
You should be able to trust that no error returned from `Create` means the puppy was indeed created.

nicko-lee replied:
Removed

anzboi commented on 10_rest/nickolee/cmd/puppy-server/main.go:
```diff
@@ -0,0 +1,122 @@
+	for _, pup := range puppies {
+		pup := pup // to avoid scopelint error: Using a reference for the variable on range scope `pup`
+		createdPuppy, _ := phs.Storage.CreatePuppy(&pup)
+		fmt.Fprintf(out, "Puppy with ID %d has been created\n", createdPuppy)
```
I like this, effectively logging successful loading of each puppy.

nicko-lee replied:
👍

anzboi commented on 10_rest/nickolee/cmd/puppy-server/main.go:
```diff
@@ -0,0 +1,122 @@
+
+	rest.SetupRoutes(r, *phs)
+
+	fmt.Printf("Starting server on port %s. Try:\n", addr)
```
This line is good, I wouldn't bother with the other lines though.

nicko-lee replied:
no problem have removed this. Just was following this (second) example from the course slides 😂
https://go-course.org/gocourse-2019-05-melbourne.slide#104

anzboi commented on 10_rest/nickolee/cmd/puppy-server/main.go:
```diff
@@ -0,0 +1,122 @@
+	rest.SetupRoutes(r, *phs)
+
+	fmt.Printf("Starting server on port %s. Try:\n", addr)
+	fmt.Printf("  curl localhost%s/api/puppy -d ", addr)
```
You don't normally see output like this when you start a server. Consider the following.
1. What if it's not running on localhost
2. What if you didn't provide a datafile (getting puppy 4 would fail)

Also, you are posting a puppy, so it needs `curl -X POST stuff`, though I wouldn't keep it.

nicko-lee replied:
ok will delete this perhaps move to the readme

anzboi commented on 10_rest/nickolee/pkg/puppy/rest/rest.go:
```diff
@@ -0,0 +1,130 @@
+)
+
+const (
+	invalidInputMsg = "Invalid input, ensure ID and JSON are valid"
```
This error message is used both when ID and JSON are invalid. Think about how you might separate these two error cases out to give the consumer a more accurate error message.

nicko-lee replied:
fixed

anzboi commented on 10_rest/nickolee/pkg/puppy/rest/rest.go:
```diff
@@ -0,0 +1,130 @@
+
+const (
+	invalidInputMsg = "Invalid input, ensure ID and JSON are valid"
+	notFoundMsg     = "Not Found: Apologies, for the puppy you seek does not exist"
```
`Apologies`
 so polite 👍

nicko-lee replied:
haha :)

anzboi commented on 10_rest/nickolee/pkg/puppy/rest/rest.go:
```diff
@@ -0,0 +1,130 @@
+
+// SetupRoutes maps the exposed endpoints with their respective handlers
+func SetupRoutes(r chi.Router, phs PuppyHandlerAndStorer) {
+	r.Get("/api/puppy/{id}", phs.handleGet)
```
I would suggest experimenting with subroutes
```
r.Route("/path", func(r chi.Router) {
  r.Get("/subpath", handler)
})
```

Also, usually API's served over http have a version in the route, `api/v1/path`

nicko-lee replied:
Good suggestions. Didn't know you could put routers in a router. That's pretty neat. Done!

anzboi commented on 10_rest/nickolee/pkg/puppy/rest/rest_test.go:
```diff
@@ -0,0 +1,205 @@
+package rest
```
This file is too similar to @willshen8. `rest.go` was inevitably going to look similar due to you both using `chi`, but testing could go many different ways.

Suggestion: split the one `TestRestAPI` into four, one for each method (eg: `TestGetPuppy`). This will also simplify the test data since each test case doesn't need a method (you'll know what it is based on the function you are in).

You can also look at using `httptest.Server` instead of `Recorder`, see how that changes your tests.

nicko-lee replied:
Did what you suggested. Rewrote test cases. Broke them up into four. No longer using `Recorder` and using `httptest.Server` instead which did indeed change my tests significantly as you predicted

anzboi reviewed (changes requested):
Unfortunately I have to insist on some changes to `rest_test.go`, they are too similar to @willshen8 and I think you will benefit from doing them again, I've suggested how you might go about this.

Other than that, left a lot of other comments on how you might change things a little just to clean it up a bit

nicko-lee committed:
Update 10_rest/nickolee/README.md

Co-Authored-By: Daniel Cantos <43394287+anzboi@users.noreply.github.com>

Commit:
Fix changes from PR comments

Main work being rewriting all API tests

anzboi commented on 10_rest/nickolee/pkg/puppy/rest/rest_test.go:
```diff
@@ -0,0 +1,291 @@
+}
+
+// a setup function that returns a preconfigured router that our test client can use
+func createTestRouter() *chi.Mux {
```
Pass in a `*testing.T` object and use `require.NoError(t, err)` instead of panicking on errors.

This will give you the nice test output for errors in test setup, as well as the tests itself.

anzboi commented on 10_rest/nickolee/pkg/puppy/rest/rest.go:
```diff
@@ -0,0 +1,148 @@
+	return &PuppyHandlerAndStorer{Storage: s}
+}
+
+// Implementing 1/4 methods in PuppyHandler interface: puppy handler for GET /api/puppy/{id}
```
`PuppyHandler` interface no longer exists, need to update this comment

nicko-lee replied:
Fixed

anzboi commented on 10_rest/nickolee/pkg/puppy/rest/rest.go:
```diff
@@ -0,0 +1,148 @@
+	})
+}
+
+// func (r *Rest) Handle(w http.ResponseWriter, r *http.Request) {
```
Removed commented code

nicko-lee replied:
Removed

anzboi commented on 10_rest/nickolee/pkg/puppy/rest/rest.go:
```diff
@@ -0,0 +1,148 @@
+package rest
```
One more change I would like you to consider, having a `errorHandler` method that checks if a returned error is a `puppy.Error` and set response status and message based on that.

For example, currently a get calls `Store.Read`, which may return an error. You assume that error to mean 404 Not Found, but it could be 500 (think about other possible implementation of store, ie: sql database).

in other words, have an error handler function specifically for handling errors from the storer
```
func handlerStoreError(w http.ResponseWriter, err error) {
  switch e := err.(type) {
  case *puppy.Error:
    // use puppy error to define the response status and body
  default:
    // handle otherwise (probably a 500)
  }
}
```

You only need to use this function for errors returned from the store, for other error cases you can handle them as you are right now with `http.Error`

anzboi reviewed (changes requested):
Good work nicko, looking much better.

Still have a few more comments to clean it up a little more

Commit:
Fix further PR comments

Mainly around error handling for the puppy handler

Commit:
Fix lint error

Oddly enough when I lint locally this does not show up at all

anzboi commented:
LGTM!
well done on completing 10 labs

nicko-lee committed:
Merge branch 'master' into lab10
